backend/fms_core/template_importer/row_handlers/library_capture/library_capture.py

```diff
@@ -27,9 +27,14 @@ def process_row_inner(self, capture_batch_info, source_sample, volume_used,
             self.errors['volume_used'] = f"Volume used ({volume_used}) exceeds the current volume of the sample ({source_sample_obj.volume})"
 
         if source_sample_obj:
+            # Add a warning if the library has failed qc
+            if any([source_sample_obj.quality_flag is False, source_sample_obj.quantity_flag is False]):
+                self.warnings["qc_flags"] = (f"Source library {source_sample_obj.name} has failed QC.")
+
             # Check if sample is not a library or a pool of libraries
             if not source_sample_obj.is_library:
                 self.errors['source_sample'] = f"Source sample {source_sample_obj.name} must be a library to be captured."
+
             # Make sure no Capture were previously done on the library
             error = []
             for derived_sample in source_sample_obj.derived_samples.all():
```

Comment:
Same problem here. If sample is not a library next line will check `derived_sample.None.library_selection` and will crash.

Reply:
There won't be non-library coming in here. The issue i have is that we can test every single step of the way if we have a library. With a layered architecture like ours it adds overhead and lines of codes but do not really provide much value. The service layers should have all user input validated beforehand if they can be. even if i add a check here it will not get back to the user ... In the end it's up to us not to call this service with something that is not a library. Its not a valid option to let this not crash by returning an error. Its safer if it crashes. Maybe we should start using some assert statements.